native/python/src/fairseq2n/bindings/data/data_pipeline.cc

```diff
@@ -121,9 +121,10 @@ data_pipeline_tracker::delete_alive_pipelines()
     for (auto &weakref : alive_pipelines_) {
         py::object pipeline_obj = weakref();
 
+        // If the pipeline has a cyclic reference during interpreter shutdown,
+        // GC skips calling its weakref callback; so this check can be true.
         if (pipeline_obj.is_none())
-            throw_<internal_error>(
-                "One of the tracked data pipelines has already been deleted. Please file a bug report.");
+            continue;
```

Comment:
would it be useful to keep some prints here ?

Reply:
Thought about that, but couldn't decide whether it is sensible. Maybe we can have some warning log? I want to avoid being extra noisy since a cyclic reference does not necessarily mean a leak. Python GC typically correctly disposes them, so the user might already be aware of them. Still can add a TODO for having a warning log (once we have the C++ logging API ready)